sycl/doc/OptionalDeviceFeatures.md

```diff
@@ -0,0 +1,502 @@
+# Behavior for optional kernel features
+
+This design document describes the changes that are needed in DPC++ in order to
+conform to the SYCL 2020 specification regarding the behavior of applications
+that use optional kernel features.  An optional kernel feature is any feature
+that is supported by some devices and not by others.  For example, not all
+devices support 16-bit floating point operations, so the `sycl::half` data type
+is an optional kernel feature.  Some DPC++ extensions like AMX are also
+optional kernel features.
+
+The requirements for this design come mostly from the SYCL 2020 specification
+[section 5.7 "Optional kernel features"][1] but they also encompass the C++
+attribute `[[sycl::requires()]]` that is described in [section 5.8.1 "Kernel
+attributes"][2] and [section 5.8.2 "Device function attributes"][3].
+
+[1]: <https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:optional-kernel-features>
+[2]: <https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:kernel.attributes>
+[3]: <https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#_device_function_attributes>
+
+
+## Requirements
+
+There are several categories of requirements covered by this design:
+
+* The front-end compiler must issue a diagnostic in some cases when a kernel
+  uses an optional feature.  However, the front-end compiler must **not**
+  generate a diagnostic in other cases.
+
+* The runtime must raise an exception when a kernel using optional features
+  is submitted to a device that does not support those features.  This
+  exception must be raised synchronously from the kernel invocation command
+  (e.g. `parallel_for()`).
+
```

Comment:
Behavior of the special case - kernel_bundle::compile() and link() - needs to be described too. Maybe in 2020 spec too.